### PR TITLE
ci: run s3tests

### PR DESCRIPTION
**Pull request by m-ildefons**
Run s3 test suite

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: .github/workflows/build-test-radosgw.yml [hunk 1]
```diff
@@ -43,6 +47,25 @@ jobs:
             -e WITH_TESTS=ON \
             quay.io/s3gw/build-radosgw:latest
 
+      - name: Pack Build artifacts
+        run: |
+          tar -czvf build-results.tar.gz \
+            ceph/build/bin/radosgw \
+            ceph/build/lib/libceph-common.so \
+            ceph/build/lib/libceph-common.so.2 \
+            ceph/build/lib/libradosgw.so \
```
Comment (jecluis) [inline .github/workflows/build-test-radosgw.yml]: haven't the `libradosgw.so*` files been removed?

**Review by jecluis — CHANGES_REQUESTED** [review]
looks good except for the `libradosgw.so` files being packaged.

**Review by jecluis — DISMISSED** [review]

**Review by jecluis — DISMISSED** [review]

File: .github/workflows/build-test-radosgw.yml [hunk 2]
```diff
@@ -42,6 +46,22 @@ jobs:
...
+        run: |
+          tar -czvf build-results.tar.gz \
+            ceph/build/bin/radosgw \
+            ceph/build/lib/libceph-common.so \
+            ceph/build/lib/libceph-common.so.2 \
+            ceph/build/lib/librados.so \
+            ceph/build/lib/librados.so.2 \
+            ceph/build/lib/librados.so.2.0.0
+
+      - name: Save radosgw binary
```
Comment (irq0) [inline .github/workflows/build-test-radosgw.yml:59]: Nit: plural or artifacts

File: .github/workflows/build-test-radosgw.yml [hunk 3]
```diff
@@ -113,3 +133,98 @@ jobs:
           ./sfs-smoke-test.sh 127.0.0.1:7480
 
           docker kill $CONTAINER
+
+  run-s3gw-tests:
+    strategy:
+      matrix:
+        op:
+          - create
+          # - list
```
Comment (irq0) [inline .github/workflows/build-test-radosgw.yml]: Nit: comment on why this is commented out would be nice

File: .github/workflows/build-test-radosgw.yml [hunk 4]
```diff
@@ -113,3 +133,98 @@ jobs:
...
+
+      - name: Install dependencies
+        run: |
+          YQ_GH_URL=https://github.com/mikefarah/yq
+          sudo apt-get update
+          sudo apt-get install -y \
+            python3-nose \
+            wget
+
+          # Unfortunately, since yq is only available through snap on Ubunut and
```
Comment (irq0) [inline .github/workflows/build-test-radosgw.yml]: Ubuntu

File: .github/workflows/build-test-radosgw.yml [hunk 5]
```diff
@@ -113,3 +133,98 @@ jobs:
...
+          export CEPH_DIR=${GITHUB_WORKSPACE}/ceph
+          export OUTPUT_DIR=${GITHUB_WORKSPACE}/s3test-results
+          export S3TEST_REPO=${GITHUB_WORKSPACE}/s3tests
+          export FORCE_CONTAINER=ON
+
+          export FIXTURES=$GITHUB_WORKSPACE/ceph/qa/rgw/store/sfs/tests/fixtures
+          export S3TEST_CONF=$FIXTURES/s3tests.conf
+
+          pushd s3tests
+          for t in $(yq \
```
Comment (irq0) [inline .github/workflows/build-test-radosgw.yml]: the yq dependency seems more trouble than its worth, especially with the rather simple 'get map key, output as space sep list' usage.

Consider python instead:

```
python <<EOF
from yaml import load
print(" ".join(load(open('s3-tests.yaml'))["bucket_create"]))
EOF

```

use either with for t in ..
or better use `\n` and pipe to while read t; do ..

File: .github/workflows/build-test-radosgw.yml [hunk 6]
```diff
@@ -113,3 +133,98 @@ jobs:
...
+          sudo apt-get update
+          sudo apt-get install -y \
+            python3-nose \
+            wget
+
+          # Unfortunately, since yq is only available through snap on Ubunut and
+          # that doesn't work in docker containers (at least not out of the
+          # box), this abomination is the way to go to install yq.
+          sudo wget \
+            "${YQ_GH_URL}/releases/latest/download/yq_linux_amd64" \
```
Comment (irq0) [inline .github/workflows/build-test-radosgw.yml]: This is basically code execution in our build workers by the yq project.
latest may point to anything. For basic security we may want to pin to version and do a checksum check. Better: replace with python (see below)

File: .github/workflows/build-test-radosgw.yml [hunk 7]
```diff
@@ -113,3 +133,98 @@ jobs:
...
+            sudo chmod +x /usr/bin/yq
+
+          python3 -m pip \
+            install --upgrade -r s3tests/requirements.txt
+
+      - name: Run S3tests ${{ matrix.op }} ${{ matrix.resource }}
+        run: |
+          set -x
+
+          export CEPH_DIR=${GITHUB_WORKSPACE}/ceph
```
Comment (irq0) [inline .github/workflows/build-test-radosgw.yml]: Needs quoting. This line and below

**Review by irq0 — CHANGES_REQUESTED** [review]

File: .github/workflows/build-test-radosgw.yml (hunk 5 above)
Comment (jecluis) [reply .github/workflows/build-test-radosgw.yml]: this also presumes 'PyYaml' as an installed dependency though. I agree with not installing "random" files from the internet, but doing it with Python will require the dependency to be installed.

**Review by jecluis — CHANGES_REQUESTED** [review]
LGTM, but I agree with @irq0 that we shouldn't be downloading "random" execution code to our runners.

**Comment by m-ildefons** [comment]
I simplified this quite a bit:
- Removed the build matrix. It necessitated to split the tests into groups in advance, which wasn't as simple as just "run all s3tests except these ones that are known to fail". This costs us a bit of parallelism in running tests but also greatly simplifies results collection, which is important.
- Pinned `yq` in place and added checksum checking
- Rebased onto `s3gw` branch
- Quoting and other improvements (typos) to the shell code in the workflow

File: .github/workflows/build-test-radosgw.yml (hunk 2 above)
Comment (m-ildefons) [reply .github/workflows/build-test-radosgw.yml:59]: Not sure what you mean?

File: .github/workflows/build-test-radosgw.yml (hunk 3 above)
Comment (m-ildefons) [reply .github/workflows/build-test-radosgw.yml]: I removed the build matrix for other reasons. This was commented out due to me not yet having gone through the listing tests and curated them by passing/failing.

File: .github/workflows/build-test-radosgw.yml (hunk 4 above)
Comment (m-ildefons) [reply .github/workflows/build-test-radosgw.yml]: Ack.

**Review by jecluis — DISMISSED** [review]

File: .github/workflows/build-test-radosgw.yml (hunk 5 above)
Comment (m-ildefons) [reply .github/workflows/build-test-radosgw.yml]: `yq` is needed for the s3tests-runner script anyways, as it is what I use to compile an aggregated report about which tests pass and which fail. I'm not keen on re-implementing what I need from it in python, especially not the more complex parts like inserting into maps and converting to JSON. That being said, I pinned the version down with a checksum check for good measure.

File: .github/workflows/build-test-radosgw.yml (hunk 7 above)
Comment (m-ildefons) [reply .github/workflows/build-test-radosgw.yml]: Done

File: .github/workflows/build-test-radosgw.yml (hunk 6 above)
Comment (m-ildefons) [reply .github/workflows/build-test-radosgw.yml]: Ack. See above

File: .github/workflows/build-test-radosgw.yml (hunk 2 above)
Comment (irq0) [reply .github/workflows/build-test-radosgw.yml:59]: name doesn't do what it says

File: .github/workflows/build-test-radosgw.yml (hunk 5 above)
Comment (irq0) [reply .github/workflows/build-test-radosgw.yml]: :+1:

**Review by irq0 — APPROVED** [review]
lgtm

**Pull request merged**